nx/guides/advanced/complex_fft.livemd

```diff
@@ -0,0 +1,297 @@
+
+```elixir
+one = Complex.new(1,1)
+```
+
+Its norm is $ \sqrt{2}$, and angle is $\pi/4$.
+
+```elixir
+sqrt_2 = :math.sqrt(2)
+pi_4 = :math.pi()/4
+
+{
+  Complex.abs(one) == sqrt_2,
+  Complex.phase(one) == pi_4
+}
+```
+
+Its polar form is:
+
+$ \sqrt{2}\big(\cos(\pi/4) + i\sin(\pi/4)\big) = \sqrt{2} \exp^{i \pi/4}$.
+
+We can use `Complex.from_polar/2`:
+
+```elixir
+cx_one = Complex.from_polar(sqrt_2, pi_4)
+```
+
+If we need a tensor, we would do:
+
+```elixir
+Nx.tensor(one)
+```
+
+We can use directly `Nx` to build a complex from its cartesian coordinates:
+
+```elixir
+t_one = Nx.complex(1,1)
+```
+
+and compute its norm and phase:
+
+```elixir
+{
+  Nx.abs(t_one),
+  Nx.phase(t_one)
+}
+```
+
+These are "tensor" aware functions.
+
+```elixir
+t = Nx.stack([t_one,Nx.complex(1,0)])
+
+Nx.abs(t)
+```
+
+We also have the imaginary number $i$ defined within `Nx.Constants`. Once imported, we can use it:
+
+```elixir
+i = Nx.Constants.i()
+
+Nx.add(1 , i)
+```
+
+It can be useful within a numerical function, defined using `defn`.
+
+For example, we can do:
+
+```elixir
+defmodule Example do
+  import Nx.Defn
+  import Nx.Constants, only: [i: 0]
+
+  defn rotate(z) do
+     i() * z
+  end
+end
+
+Nx.equal(Example.rotate(one), Nx.complex(-1,1))
+```
+
+## Discrete Fourrier Transform
+
+Suppose we have a periodic signal, and we want to sample it to understand the signal.
+
+We will run a Fast Fourrier Transform. It transforms a time-domain function into the frequency domain. Let's see how.
+
+<https://en.wikipedia.org/wiki/Discrete_Fourier_transform>
+
+<!-- livebook:{"break_markdown":true} -->
+
+### Build the signal
+
+Let's build a signal that we want to understand.
+
+It will be the sum of two sinusoidal signals, one at 5Hz, and one at 20Hz with the corresponding amplitudes (1, 0.5).
+
+$f(t) = \sin(2\pi*5*t) + \frac12 \sin(2\pi*20*t)$
+
+Suppose we can sample at `fs=50Hz` (meaning 50 samples per second) and our aquisition time is `duration = 1s`.
+
+We build a time series of `fs` points equally spaced with the given `duration` interval with `Nx.linspace`.
+
+For each value of this serie (the discrete time $t$), we will calculate the "hidden" signal $f(t)$ in the module below. This will give us the sampling.
+
+```elixir
+defmodule Signal do
+  import Nx.Defn
+  import Nx.Constants, only: [pi: 0]
+
+  defn source(t) do
+    f1 = 5; f2 = 20;
+    Nx.sin(2 * pi() * f1 * t ) + 1/2 * Nx.sin(2 * pi() * f2 * t)
+  end
+
+  defn sample(opts) do
+    fs = opts[:fs]
+    duration = opts[:duration]
+    sampling = Nx.linspace(0, duration, n: fs, endpoint: false, type: {:f, 32})
+    source(sampling)
+  end
+end
+```
+
+We sample our signal at fs=50Hz during 1s:
+
+```elixir
+fs = 50; duration= 1
+
+sample = Signal.sample(fs: fs, duration: 1)
+```
+
+### Analyse the signal with DFT
+
+The DFT algorithm will **approximates the original signal**. It gives us the following information:
```

Comment:
```suggestion
Because our signal contains many periods of the underlying function, the DFT results will contain some noise. This noise stem both from the fact that we're likely cutting of the signal before it repeats, and from the fact that we have a specific frequency resolution which ends up grouping our individual components into frequency bins - the latter isn't really a problem as we have chosen `fs` to be [fast enough](https://en.wikipedia.org/wiki/Nyquist%E2%80%93Shannon_sampling_theorem).
```